consensus/scp/src/quorum_set.rs

```diff
@@ -39,18 +29,68 @@ pub enum QuorumSetMember<ID: GenericNodeId> {
     InnerSet(QuorumSet<ID>),
 }
 
-impl<
-        ID: GenericNodeId
-            + Clone
-            + Debug
-            + Display
-            + Serialize
-            + DeserializeOwned
-            + Eq
-            + PartialEq
-            + Hash,
-    > QuorumSet<ID>
-{
+impl<ID: GenericNodeId> PartialEq for QuorumSetMember<ID> {
+    fn eq(&self, other: &QuorumSetMember<ID>) -> bool {
+        match self {
+            QuorumSetMember::Node(self_node) => match other {
+                QuorumSetMember::Node(other_node) => self_node == other_node,
+                _ => false,
+            },
+            QuorumSetMember::InnerSet(self_qs) => match other {
+                QuorumSetMember::InnerSet(other_qs) => self_qs == other_qs,
+                _ => false,
+            },
+        }
+    }
+}
+impl<ID: GenericNodeId> Eq for QuorumSetMember<ID> {}
+
+impl<ID: GenericNodeId> Hash for QuorumSetMember<ID> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            QuorumSetMember::Node(self_node) => self_node.hash(state),
+            QuorumSetMember::InnerSet(self_qs) => self_qs.hash(state),
+        }
+    }
+}
+
+/// The quorum set defining the trusted set of peers.
+#[derive(Clone, Debug, Ord, PartialOrd, Serialize, Deserialize, Digestible)]
+pub struct QuorumSet<ID: GenericNodeId = NodeID> {
+    /// Threshold (how many members do we need to reach quorum).
+    pub threshold: u32,
+
+    /// Members.
+    pub members: Vec<QuorumSetMember<ID>>,
+}
+
+impl<ID: GenericNodeId> PartialEq for QuorumSet<ID> {
+    fn eq(&self, other: &QuorumSet<ID>) -> bool {
+        if self.threshold == other.threshold && self.members.len() == other.members.len() {
+            // sort before comparing
+            let mut self_members: Vec<QuorumSetMember<ID>> = self.members.clone();
+            let mut other_members: Vec<QuorumSetMember<ID>> = other.members.clone();
+            self_members.sort();
+            other_members.sort();
+            return self_members == other_members;
+        }
+        false
+    }
+}
+impl<ID: GenericNodeId> Eq for QuorumSet<ID> {}
+
+impl<ID: GenericNodeId> Hash for QuorumSet<ID> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.threshold.hash(state);
```

Comment:
@jcape / @mfaulk  do we add field and domain separators for `Hash` implementations? I think we do?

Reply:
It's not a cryptographic hash, so it's less important, but it's generally good practice to add the name of the object.

Reply:
easy to add but I don't think derived Hash uses separators...